### PR TITLE
Object Store Data Integrity

### PR DESCRIPTION
**Pull request by toly-kournik**
CheckpointMsg to include state (blockId).
State Transfer to return blockid with a state digest.
Don't use tableOfStableCheckpoints in ROR, but rather use MsgCertificate
in CheckpointInfo.

### CODE REVIEW

File: bftengine/src/bftengine/ReadOnlyReplica.cpp [hunk 1]
```diff
@@ -102,58 +102,32 @@ void ReadOnlyReplica::onMessage<CheckpointMsg>(CheckpointMsg *msg) {
...
+      KVLOG(msg->senderId(), msgGenReplicaId, msgSeqNum, msgEpochNum, msg->size(), msg->isStableState(), msg->state())
+          << ", state digest: " << msg->digestOfState().toString());
   // Reconfiguration cmd block is synced to RO replica via reserved pages
   auto epochNumberFromResPages = ReconfigurationCmd::instance().getReconfigurationCommandEpochNumber();
   if (epochNumberFromResPages.has_value()) {
     replicasLastKnownEpochVal = epochNumberFromResPages.value();
   }
   // not relevant
-  if (!msgIsStable || msgSeqNum <= lastExecutedSeqNum) return;
+  if (!msg->isStableState() || msgSeqNum <= lastExecutedSeqNum || msgEpochNum != replicasLastKnownEpochVal) return;
```
Comment (yontyon) [inline bftengine/src/bftengine/ReadOnlyReplica.cpp:117]: Epoch number is not relevant for read-only replica

File: bftengine/src/bftengine/messages/CheckpointMsg.hpp [hunk 2]
```diff
@@ -47,6 +51,10 @@ class CheckpointMsg : public MessageBase {
   void sign();
 
   void setSenderId(NodeIdType id) { sender_ = id; }
+  static bool equivalent(CheckpointMsg* a, CheckpointMsg* b) {
```
Comment (yontyon) [inline bftengine/src/bftengine/messages/CheckpointMsg.hpp:54]: What about the epoch number?

**Review by yontyon — APPROVED** [review]
LGTM!
I have only two comments about epochs

File: bftengine/src/bftengine/messages/CheckpointMsg.hpp (hunk 2 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/messages/CheckpointMsg.hpp:54]: you are right, I added only what I added :-)

File: bftengine/src/bftengine/ReadOnlyReplica.cpp (hunk 1 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/ReadOnlyReplica.cpp:117]: But still there was the epoch related code - look on the left side

**Pull request merged**